src/eduid_userdb/element.py

```diff
@@ -452,45 +458,53 @@ def add(self, element):
         if self.find(element.key):
             raise DuplicateElementViolation("Element {!s} already in list".format(element.key))
 
-        _old_list = self._elements
+        old_list = self._elements
         ElementList.add(self, element)
-
-        try:
-            assert self.primary is not None
-        except PrimaryElementViolation:
-            self._elements = _old_list
-            raise
+        self._check_primary(old_list)
         return self
 
     def remove(self, key):
         """
         Remove an existing Element from the list.
 
-        Raises PrimaryElementViolation if the operation results in != 1 primary
-        element in the list.
+        Raises PrimaryElementViolation if the operation results in 0 primary
+        element in the list but there are confirmed elements.
 
         :param key: Key of element to remove
         :type key: str | unicode
         :return: ElementList
         """
-        _old_list = self._elements
+        old_list = self._elements
         ElementList.remove(self, key)
+        self._check_primary(old_list)
+        return self
+
+    def _check_primary(self, old_list):
+        """
+        If there are confirmed elements, there must be exactly one primary
+        element. If there are no confirmed elements, there must be 0 primary
+        elements.
 
+        :param old_list: list of elements to get back to if the constraints are violated
+        :type old_list: list
+        """
         try:
-            if self._elements:
+            if self.verified.count > 0:
                 assert self.primary is not None
-        except PrimaryElementViolation:
-            self._elements = _old_list
+            else:
+                assert self.primary is None
+        except (PrimaryElementViolation, AssertionError):
```

Comment:
Theoretically, assertions can be disabled in the Python interpreter at run-time. Better to do

if self.primary is not None:
    raise PrimaryElementViolation(...)